static/js/timerender.ts

```diff
@@ -24,6 +24,91 @@ export function clear_for_testing(): void {
     next_timerender_id = 0;
 }
 
+type DateFormat = "short" | "medium" | "long" | "day" | "full";
+type DateWithTimeFormat = "short12" | "short24" | "medium12" | "medium24" | "long12" | "long24";
+type TimeFormat = "hour12" | "hour24";
+
+type DateOrTimeFormat = DateFormat | TimeFormat | DateWithTimeFormat;
+
+// The function takes in an optional type parameter and returns
+// an options object that can be used to format a date/time using the Intl.DateTimeFormat API.
+// This function allows easy and consistent formatting of dates and times
+// with different options to customize the format according to the specified type
+function get_format_options_for_type(type?: DateOrTimeFormat): Intl.DateTimeFormatOptions {
+    const hour12FormatOptions: Intl.DateTimeFormatOptions = {
+        hourCycle: "h12",
+        hour: "numeric",
+        minute: "2-digit",
+    };
+    const hour24FormatOptions: Intl.DateTimeFormatOptions = {
+        hourCycle: "h23",
+        hour: "2-digit",
+        minute: "2-digit",
+    };
+
+    const dayFormatOptions: Intl.DateTimeFormatOptions = {weekday: "long"};
+    const fullFormatOptions: Intl.DateTimeFormatOptions = {dateStyle: "full"};
+
+    const shortFormatOptions: Intl.DateTimeFormatOptions = {day: "2-digit", month: "short"};
+    const mediumFormatOptions: Intl.DateTimeFormatOptions = {
+        ...shortFormatOptions,
+        year: "numeric",
+    };
+    const longFormatOptions: Intl.DateTimeFormatOptions = {
+        ...mediumFormatOptions,
+        ...hour24FormatOptions,
+        weekday: "short",
+    };
+
+    switch (type) {
+        case "day": // Wednesday
+            return dayFormatOptions;
+        case "full": // Wednesday, July 27, 2016
+            return fullFormatOptions;
+        case "hour12": // 01:30 PM
+            return hour12FormatOptions;
+        case "hour24": // 13:30
+            return hour24FormatOptions;
+        case "short": // Jul 27
+            return shortFormatOptions;
+        case "short12": // Jul 27, 01:30 PM
+            return {...shortFormatOptions, ...hour12FormatOptions};
+        case "short24": // Jul 27, 13:30
+            return {...shortFormatOptions, ...hour24FormatOptions};
+        case "medium": // Jul 27, 2016
+            return mediumFormatOptions;
+        case "medium12": // Jul 27, 2016, 01:30 PM
+            return {...mediumFormatOptions, ...hour12FormatOptions};
+        case "medium24": // Jul 27, 2016, 13:30
+            return {...mediumFormatOptions, ...hour24FormatOptions};
+        case "long": // Wed, Jul 27, 2016, 01:30 PM
+        case "long12": // Wed, Jul 27, 2016, 01:30 PM
+            return {...longFormatOptions, ...hour12FormatOptions};
+        case "long24": // Wed, Jul 27, 2016, 13:30
+            return {...longFormatOptions, ...hour24FormatOptions};
+        default: // Jul 27, 2016
+            return mediumFormatOptions;
```

Comment:
It might be better to just have the format be a mandatory parameter, rather than having a default here, for readability.

Reply:
I remembered why I did this. If there is no default return, we will get an ESLint error: `Expected to return a value at the end of function 'get_format_options_for_type' consistent-return.` ESLint doesn't know about types, so there's no great way of "fixing" consistent-return to know about this.

One way to fix this would be to add a variable and assign a value to it in each case, and then return it at the end. But I don't like that approach because it would require many `break` statements in each case. I would prefer to stick with the default construction at the end of switch. What do you think, @timabbott?

Reply:
Can't we just `throw new Error` in the default case?

Reply:
In any case, even if we had a default, I think every caller should definitely pass the format explicitly. The goal here is that if you're reading one of the functions that call this, which are usually complex conditionals with other options like "Yesterday" as possible outputs, you can see a clear hint of what format we're asking for.

But if we want to enforce that pattern, it's best to just make it a mandatory parameter.

Reply:
I am completely agree with you. With such a large codebase and a lot of developers, it's understandable that not everyone knows what is default value. I should have taken that into consideration.